File: src/main/java/ru/randomsoftware/java2d/simulations/LineCellularAutomate.java
package ru.randomsoftware.java2d.simulations;

import ru.randomsoftware.java2d.generators.LineGenerator;
import ru.randomsoftware.java2d.rules.Rule;

import java.util.ArrayList;
import java.util.List;

public class LineCellularAutomate {

    private final List<Boolean> initialRow;
    private final Rule<Boolean> rule;

    public LineCellularAutomate(LineGenerator<Boolean> generator, Rule<Boolean> rule, int size) {
        this.initialRow = generator.generate(size);
        this.rule = rule;
    }

    public List<List<Boolean>> buildGenerations(int n) {
        List<List<Boolean>> matrix = new ArrayList<>();
        List<Boolean> row = initialRow;
        for (int y = 0; y < n; y++) {
            matrix.add(row);
            row = calculateNextRow(row);
        }
        return matrix;
    }

    private List<Boolean> calculateNextRow(List<Boolean> row) {
        List<Boolean> nextRow = new ArrayList<>();
        nextRow.add(rule.calculate(row.get(row.size() - 1), row.get(0), row.get(1)));
        for (int i = 1; i < row.size() - 1; i++) {
            nextRow.add(rule.calculate(row.get(i - 1), row.get(i), row.get(i + 1)));
        }
        nextRow.add(rule.calculate(row.get(row.size() - 2), row.get(row.size() - 1), row.get(0)));
        return nextRow;
    }

}

File: src/main/kotlin/ru/randomsoftware/java2d/simulations/LineCellularAutomate.kt
package ru.randomsoftware.java2d.simulations

import ru.randomsoftware.java2d.generators.LineGenerator
import ru.randomsoftware.java2d.rules.Rule

class LineCellularAutomate(
        generator: LineGenerator<Boolean>,
        private val rule: Rule<Boolean>,
        size: Int
): Automate<Boolean> {

    private val initialRow: List<Boolean> = generator.generate(size)

    override fun buildGenerations(n: Int): List<List<Boolean>> {
        return generateSequence(initialRow) { calculateNextRow(it) }.take(n).toList()
    }

    private fun calculateNextRow(row: List<Boolean>): List<Boolean> {
        return row.takeLast(1).plus(row.dropLast(1))
                .zip(row)
                .zip(row.drop(1).plus(row.take(1)))
                .map {
                    Triple(it.first.first, it.first.second, it.second)
                }.map {
                    rule.calculate(it.first, it.second, it.third)
                }
    }
}